### DISCUSSION TITLE
Jamulus feedback protection, possible to turn off?

### DISCUSSION THREAD

**Discussion by harmonk** [jamulussoftware/jamulus#2066] [Technical Support & Help] [answered]
Hi All,
Using Jamulus and like it.
One thing i would see changed: Whenever there is a loud sudden noise, bump the mic for example, the feedback protection kicks in, i get a popup window and everything is muted. So three clicks: wondow close, 'mute all', my channel mute. Every time. 
I don't need this, so i wonder if there is any way to turn this off? I can't find any way.
This is not the same as 'feedback protection' in the advanced setting. That is already turned off.
Thanks for any thoughts about this.

**Comment by npostavs** [comment]
> This is not the same as 'feedback protection' in the advanced setting. That is already turned off.

If it's not that, then it sounds like a bug. What does the popup window say?

**Reply by harmonk** [reply]
Thanks for your reply. 
I can't remember precisely what it said.
I have not used this particular setup again since i wrote. As soon as i am back on, and encounter this message, i will screenshot it and post again.

**Reply by harmonk** [reply]
Ok, now i have it. Here is a screen shot.
![jamulus mute](https://user-images.githubusercontent.com/8872185/138504363-04544ea2-ba33-4f2c-a580-bfff7e665044.jpg)

**Reply by harmonk** [reply]
Again, this is not after feedback, just after any sudden loud sound such as bumping the mic.

**Reply by npostavs** [reply]
That comes from here: https://github.com/jamulussoftware/jamulus/blob/r3_8_1rc2/src/clientdlg.cpp#L1047

Looks like there is `CClientDlg::bDetectFeedback` and `CClientSettings::bEnableFeedbackDetection` which means they can contradict each other. Not sure how it's meant to work. @JohannesBrx: any ideas?

**Reply by JohannesBrx** [reply]
@npostavs:
When feedback detection is enabled by `CClientSettings::bEnableFeedbackDetection`, `CClientDlg::bDetectFeedback` is set `true` for 3 seconds after connecting and deactivated (set to `false`) by the timer `TimerDetectFeedback`. The intention behind this fairly simple feedback detection is that feedback typically occurs in the first seconds after connecting, during which time no loud noise is normally heard.

`CClientSettings::bEnableFeedbackDetection` is the user setting for enabling this feature.

**Reply by npostavs** [reply]
It looks like @harmonk's screenshot shows `bEnableFeedbackDetection` false (since the checkbox is unchecked), but still `bDetectFeedback` was set to `true` (because the popup is invoked inside the if statement). Should that be possible?

**Reply by JohannesBrx** [reply]
Actually this should not be possible. At both places where `bDetectFeedback` is set to `true`, the timer for setting it later to `false` is also started:
```  
        TimerDetectFeedback.start ( DETECT_FEEDBACK_TIME_MS );
        bDetectFeedback = true;
```

Maybe there is an issue with the timers for some systems?

To quickly fix this, we could additionally (and redundantly) check `bEnableFeedbackDetection` inside `CClientDlg::OnTimerSigMet`.

**Reply by npostavs** [reply]
Maybe something similar to #1791, e.g., the timer thread just fails to run on macOS for long periods of time?

> To quickly fix this, we could additionally (and redundantly) check `bEnableFeedbackDetection` inside `CClientDlg::OnTimerSigMet`.

That sounds sensible to me.

**Reply by JohannesBrx** [reply]
Shall I open up a new merge request or how do you typically handle a small change like this?

**Reply by pljones** [reply]
Same process -- in your fork, branch for the fix, make the change and push, then raise the pull request for review.

**Reply by npostavs** [reply]
@JohannesBrx I had the idea this morning that instead of `bDetectFeedback` being a boolean and turning it off with a timer, it could be a timestamp and only be considered enabled if it's within 3 seconds of current time. Then you wouldn't need the extra timer at all.

**Reply by pljones** [reply]
Better to store "three seconds from now" at connect and then bypass the check if "now is greater than cutoff".  That stops you doing the calculation each time.  The bool check still is simpler, of course.

**Reply by pljones** [reply]
One thing I've wondered is whether "connect" is being managed properly.  Is there any way to check whether the client "thinks it's connecting" more often than one would imagine?  (I've not looked at the code -- if this _only_ happens when the user deliberately connects, it shouldn't be an issue but if it happens when the connection succeeds, then that might be being triggered in unexpected ways.)

**Reply by JohannesBrx** [reply]
I opened PR #2096 with a quick fix which redundantly checks the feedback detection enable state.

@npostavs: I did not yet have spare time to implement the solution with the timestamp.

**Comment by JacquesFreud** [comment]
Since about a week on some (I think not all, but I am not sure) servers (i.e. Die Insel, Genre 1 or MTKS.F, genre 2) feedback protection kicks in way too fast. A sudden laugh, a handclap, a guitar that starts soloing, a keyboard, vocals - many sounds that were normally accepted before, all of a sudden lead to muting.
This happens both when you selected Feedback Protection in your settings or not. This has _nothing_ to do with initial startup of people who forgot to turn off their speakers or made some other mistake. It started happening about a week ago and not when this feature was introduced in version 3.8.0 
Anybody with a solution?

**Reply by npostavs** [reply]
As far I understand, the feedback detection happens only on the client side (perhaps different delay of servers could affect when the sound returns though). Are you also on macOS?

**Reply by JacquesFreud** [reply]
Yes MacOS

**Reply by softins** [reply]
> It started happening about a week ago and not when this feature was introduced in version 3.8.0 Anybody with a solution?

This is interesting. Did you change anything (even supposedly unrelated) at about that time? Such as the audio interface, or the server that you usually use?

I'd like to understand the mechanism, before implementing the sensible extra check proposed by @JohannesBrx in #2096. I can see in the code that the feedback detection timer is started:
a. In the Connect event handler.
b. When detecting that an audio device has changed.

So maybe there is the possibility that one of these is getting called repeatedly for some reason? (@pljones suggested the same in one of his comments above)

@JacquesFreud What version of macOS are you using, and what kind of audio interface? Does the issue still occur? I have a Mac too, so could try to reproduce the issue.

**Reply by softins** [reply]
OK, I think I have understood what is happening:

1. `bDetectFeedback` only gets set to `true`, and the 3-second timer started, IF `bEnableFeedbackDetection` is also `true`. Since the OP said the latter setting is already `false`, this can't be happening.
2. `bDetectFeedback` only gets set to `false` within `CClientDlg::OnTimerDetectFeedback()` and `CClientDlg::Disconnect()`. Once this has happened, the level check and subsequent popup CAN'T happen.
3. However, `bDetectFeedback` is NOT getting initialised when the `CClientDlg` object is instantiated. This means that, depending on the OS and runtime environment, it could randomly contain any value when the program is started up. UNTIL the first time the client disconnects from a server, after which it will be fine.

So this means that after starting the program, and connecting for the first time to a server, the problem may or may not occur, depending on whether the uninitialised `bDetectFeedback` happens to be zero or non-zero. After disconnecting from the server, it should then be possible to connect to any server without the problem occurring for the rest of the time the program is running. Exiting and restarting the program will allow the issue to occur again for the duration of the next connection. @harmonk please could you test and confirm this?

So the correct fix is to initialise `bDetectFeedback` to `false` in the constructor for `CClientDlg`. Then the extra check in #2096 will not be necessary, although it doesn't hurt either.

**Reply by softins** [reply]
I have just managed to reproduce the above behaviour reliably, by temporarily adding an initialisation of `bDetectFeedback` to `true` in `CClientDlg::CClientDlg()`. Then the described behaviour occurs during the first connection, but not during subsequent connections within the same run of the program.

Note that for current versions of Jamulus, if the feedback detection is ENABLED in settings, then this bug can never occur.

**Reply by softins** [reply]
> So this means that after starting the program, and connecting for the first time to a server, the problem may or may not occur, depending on whether the uninitialised `bDetectFeedback` happens to be zero or non-zero. After disconnecting from the server, it should then be possible to connect to any server without the problem occurring for the rest of the time the program is running. Exiting and restarting the program will allow the issue to occur again for the duration of the next connection. @harmonk please could you test and confirm this?

Also @JacquesFreud ?

**Reply by JacquesFreud** [reply]
Ooh sorry, there were no reactions for a while and now all of a sudden a whole lot.
Let me start out by saying that I reinstalled 3.7.0 and got rid of the feedback protection. But that is not what you tech guts are after, of course. The servers that hurt me most were Die Insel in Genre 1 and MTKS.F in genre 2. It could very well be that by changing to another server solved my problem, because I have thought quite often that the problem occurred with some servers and not on others…

**Reply by softins** [reply]
> It could very well be that by changing to another server solved my problem, because I have thought quite often that the problem occurred with some servers and not on others…

Thanks, but what solved your problem was disconnecting and then reconnecting to a server. It can only occur during the first connection after starting the program.

**Reply by JacquesFreud** [reply]
Indeed! Thanks for figuring this all out!

**Comment by ghost** [comment]
I believe there may be multiple bugs in this code as suggested above.
And also in:
/home/USER/.config/Jamulus/Jamulus.ini
Where more than one variable may be saved because of updating an old version of Jamulus.ini

**Reply by rdica** [reply]
What are the vars in the ini file that conflict now?

**Reply by rdica** [reply]
@DavidSavinkoff You never responded, what vars would conflict?

**Reply by softins** [reply]
> Where more than one variable may be saved because of updating an old version of Jamulus.ini

I think you will find this cannot occur. The `Jamulus.ini` file is written from scratch when the program exits, and will just contain the current set of variables. Old variables from previous versions might be understood for compatibility when reading the ini file at startup, but will not be written out again on exit.

**Comment by JacquesFreud** [comment]
I removed the 'jamulus.ini' file and started up Jamulus. The first handclaps I tried were quite promising, David! I will see how things develop in the upcoming days and will report back as soon as I dare to make more definitive conclusions. Thanks!

**Comment by JacquesFreud** [comment]
First conclusion: removing the jamulus.ini did not solve the problem. Changing to a different server (leaving my volume settings at the same level) did help. So the problem may be a combination of Jamulus and some servers you use it on.

**Reply by rdica** [reply]
Can you note the version of Jamulus the servers are running? You can get that info from https://explorer.jamulus.io

**Reply by JacquesFreud** [reply]
This by no means a representative measuring (only three servers I tried and looked at): one that had the feedback protection kick in too fast: 3.8.0; one that was okay: 3.6.2 and another one that failed me on the feedback: 3.6.2
So version number seems not to be the problem...

**Reply by softins** [reply]
> This by no means a representative measuring (only three servers I tried and looked at): one that had the feedback protection kick in too fast: 3.8.0; one that was okay: 3.6.2 and another one that failed me on the feedback: 3.6.2 So version number seems not to be the problem...

If there are specific servers that reliably cause the problem, could you mention one or two of them, so that we could test and reproduce the problem too?

**Reply by softins** [reply]
> > This by no means a representative measuring (only three servers I tried and looked at): one that had the feedback protection kick in too fast: 3.8.0; one that was okay: 3.6.2 and another one that failed me on the feedback: 3.6.2 So version number seems not to be the problem...
> 
> If there are specific servers that reliably cause the problem, could you mention one or two of them, so that we could test and reproduce the problem too?

Don't worry about that, but please see my reply above at https://github.com/jamulussoftware/jamulus/discussions/2066#discussioncomment-1632868

**Reply by JacquesFreud** [reply]
Die Insel in Genre 1 and MTKS.F in Genre 2. Hope this helps.

**Comment by softins** [comment, marked as answer]
This bug is fixed in #2120

**Reply by JacquesFreud** [reply]
Does this mean I can safely go back to 3.8.1 or do I need to wait for an update?

**Reply by softins** [reply]
3.8.1 does not contain the fix. The next release will, although no timeframe for the next release has been decided.

If you leave feedback detection _enabled_, the problem will never occur: the test for loudness will only happen during the first 3 seconds of a connection. So with that setting, you can safely use 3.8.1.

**Reply by JacquesFreud** [reply]
Thanks, you wonderful people!

**Comment by harmonk** [comment]
I see this only now. But thanks! So i will turn on feedback protection, and then wait for the next update which will contain the fix.

**Reply by ann0see** [reply]
Yes. The next release should be around the corner

**Reply by hoffie** [reply]
There's also a pre-release here which contains the fix already:
https://github.com/jamulussoftware/jamulus/releases/tag/r3.8.1devNightly1

(The pre-release might be missing some translations though)
